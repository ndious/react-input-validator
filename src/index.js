import Text from './text'
import Email from './email';
import Password from './password'

export default {
  Text,
  Email,
  Password,
}
