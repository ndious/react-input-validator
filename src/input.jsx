import React from 'react';

export const Input = ({
  type, name, id,
  validator, required,
  onBlur, onChange,
  ...props
}) => {

  const test = target => {
    const isEmpty = target.value.length === 0
    let isValid
    if (typeof validator === 'function') {
      isValid = validator(target.value)
    } else {
      isValid = validator.test(target.value)
    }
    const onError = required ? (isEmpty || !isValid) : !(isValid || isEmpty)

    return {
      ...target,
      isEmpty,
      isValid,
      onError,
    }
  }

  const handleChange = ev => {
    console.log('there is change', onChange)
    if (onChange) {
      return onChange({
        ...ev,
        target: test(ev.target)
      })
    }
  }

  const handleBlur = ev => {
    if (onBlur) {
      return onBlur({
        ...ev,
        target: test(ev.target)
      })
    }
  }

  return (
    <input
      type={type}
      name={name || id}
      id={id || name}
      onChange={handleChange}
      onBlur={handleBlur}
      {...props}
      required
    />
  )
}
