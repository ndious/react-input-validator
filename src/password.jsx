import React from 'react'
import Input from './input'

const Password = ({min, max, upper, lower, special, number, ...props}) => {
  let validator = '^'
  validator += lower ? '(?=.*[a-z])' : ''
  validator += upper ? '(?=.*[A-Z])' : ''
  validator += number ? '(?=.*\\d)' : ''
  validator += special ? '(?=.*[$@$!%*?&])' : ''
  validator += `[\\w\\d$@$!%*?&]{${min || 0},${max || ''}}$`

  return (
    <Input
      type="password"
      validator={new RegExp(validator)}
      {...props}
    />
  )
}

export default Password
