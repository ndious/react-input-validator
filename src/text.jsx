import React from 'react'
import {Input }from './input'

const Text = ({validate, ...props}) => (
  <Input type="text" validator={validate || /.*/} {...props} />
)

export default Text
