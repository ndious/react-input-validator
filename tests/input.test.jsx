import React from 'react'
import { shallow, mount } from 'enzyme'

import { Input } from '../src/input'

it('should render a input', () => {
    const wrapper = shallow(<Input />)
    expect(wrapper).toMatchSnapshot()
});

it('should render a input id', () => {
    const wrapper = shallow(<Input name="foo" />)
    expect(wrapper.find('#foo').length).toBe(1)
});

it('should called onChange event', () => {
  const handleChange = jest.fn()
  const wrapper = mount(<Input id="test" onChange={handleChange} />)
  console.log(wrapper.find('#test'))
  wrapper.find('#test').simulate('keyDown', { which: 'a' })
  expect(handleChange).toHaveBeenCalled()
})