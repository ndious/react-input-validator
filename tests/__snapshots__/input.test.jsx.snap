// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a input  1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <input
    id={undefined}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type={undefined}
/>,
  "nodes": Array [
    <input
      id={undefined}
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type={undefined}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Input />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
        "_debugID": 4,
        "_renderedOutput": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Input />,
}
`;

exports[`should render a input 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <input
    id={undefined}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type={undefined}
/>,
  "nodes": Array [
    <input
      id={undefined}
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type={undefined}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Input />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
        "_debugID": 2,
        "_renderedOutput": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Input />,
}
`;

exports[`should render a input 2`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <input
    id={undefined}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type={undefined}
/>,
  "nodes": Array [
    <input
      id={undefined}
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type={undefined}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Input />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
        "_debugID": 4,
        "_renderedOutput": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Input />,
}
`;

exports[`should render a input id 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <input
    id={undefined}
    name={undefined}
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    type={undefined}
/>,
  "nodes": Array [
    <input
      id={undefined}
      name={undefined}
      onBlur={[Function]}
      onChange={[Function]}
      required={true}
      type={undefined}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Input />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {},
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
        "_debugID": 4,
        "_renderedOutput": <input
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          required={true}
          type={undefined}
/>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Input />,
}
`;
